java/arcs/core/storage/CapabilitiesResolverNew.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage
+
+import arcs.core.common.ArcId
+import arcs.core.data.CapabilitiesNew
+import arcs.core.data.CollectionType
+import arcs.core.data.EntityType
+import arcs.core.data.ReferenceType
+import arcs.core.data.Schema
+import arcs.core.data.SingletonType
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+import arcs.core.type.Type
+
+/**
+ * [CapabilitiesResolverNew] is a factory class that creates [StorageKey]s based on the registered
+ * providers and given [CapabilitiesNew].
+ */
+class CapabilitiesResolverNew(
+    val options: Options,
+    factoriesList: List<StorageKeyFactory> = listOf(),
+    val selector: FactorySelector = SimpleCapabilitiesSelector()
+) {
+    val factories: Map<String, StorageKeyFactory>
+
+    init {
+        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
+            "Storage keys protocol must be unique $factoriesList."
+        }
+        factories = factoriesList.associateBy { it.protocol }.toMutableMap()
+        CapabilitiesResolverNew.defaultStorageKeyFactories.forEach { (protocol, factory) ->
+            factories.getOrPut(protocol) { factory }
+        }
+    }
```

Comment:
Use a constructor? 


```suggestion
class CapabilitiesResolverNew(
    private val options: Options,
    private val selector: FactorySelector = SimpleCapabilitiesSelector(),
    private val factories: Map<String, StorageKeyFactory>
) {

    constructor(
         options: Options, 
         selector: FactorySelector = SimpleCapabilitiesSelector(),
         factoriesList: List<StorageKeyFactory> = listOf()
    ) : this(options, selector, CapabilitiesResolverNew.getFactorMap(factoriesList)) 
    
    // This should go into the companion object
    fun getFactorMap(factoriesList: List<StorageKeyFactory>): Map<String, StorageKeyFactory> {
        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
            "Storage keys protocol must be unique $factoriesList."
        }
        val factories = factoriesList.associateBy { it.protocol }.toMutableMap()
        CapabilitiesResolverNew.defaultStorageKeyFactories.forEach { (protocol, factory) ->
            factories.getOrPut(protocol) { factory }
        }
        return factories
    }
```

Reply:
done.
one reason i can think of why this is preferable (despite being more verbose) it that the class declaration clearly shows all data members. are there other reasons?